src/vmm/src/gdb/arch/aarch64.rs

```diff
@@ -1,62 +1,296 @@
+    // asserted with our 4kb page, plus the offset of 3
+    let lower_mask: u64 = 0xFFF;
+    // A mask for a physical address mask with the lower 12 bits cleared
+    let desc_mask: u64 = pa_address_mask & !lower_mask;
+
+    let page_indicies = [
+        (gva >> 36) & lower_mask,
+        (gva >> 27) & lower_mask,
+        (gva >> 18) & lower_mask,
+        (gva >> 9) & lower_mask,
+    ];
+
+    // Transition table base register used for initial table lookup
+    // take the bottom 48 bits from the register value
+    let mut address: u64 = get_sys_reg(TTBR1_EL1, vcpu_fd)? & pa_address_mask;
+    let mut level = 0;
+
+    while level < 4 {
+        // Clear the bottom 3 bits from this address
+        let pte = read_address(vmm, (address + page_indicies[level]) & PTE_ADDRESS_MASK)?;
+        address = pte & desc_mask;
+
+        // If this is a valid table entry and we aren't at the end of the page tables
+        // then loop again and check next level
+        if (pte & 2 != 0) && (level < 3) {
+            level += 1;
+            continue;
+        }
+        break;
+    }
+
+    let page_size = 1u64 << ((9 * (4 - level)) + 3);
+    // Clear bottom bits of page size
+    address &= !(page_size - 1);
+    address |= gva & (page_size - 1);
+    Ok(address)
 }
 
 /// Configures the kvm guest debug regs to register the hardware breakpoints
 fn set_kvm_debug(
-    _control: u32,
-    _vcpu_fd: &VcpuFd,
-    _addrs: &[GuestAddress],
+    control: u32,
+    vcpu_fd: &VcpuFd,
+    addrs: &[GuestAddress],
 ) -> Result<(), GdbTargetError> {
-    unimplemented!()
+    let mut dbg = kvm_guest_debug {
+        control,
+        ..Default::default()
+    };
+
+    for (i, addr) in addrs.iter().enumerate() {
+        // DBGBCR_EL1 (Debug Breakpoint Control Registers, D13.3.2):
+        // bit 0: 1 (Enabled)
+        // bit 1~2: 0b11 (PMC = EL1/EL0)
+        // bit 5~8: 0b1111 (BAS = AArch64)
+        // others: 0
+        dbg.arch.dbg_bcr[i] = 0b1 | (0b11 << 1) | (0b1111 << 5);
+        // DBGBVR_EL1 (Debug Breakpoint Value Registers, D13.3.3):
+        // bit 2~52: VA[2:52]
+        dbg.arch.dbg_bvr[i] = (!0u64 >> 11) & addr.0;
+    }
+
+    vcpu_fd.set_guest_debug(&dbg)?;
+
+    Ok(())
+}
+
+/// Bits in a Vcpu pstate for IRQ
+const IRQ_ENABLE_FLAGS: u64 = 0x80 | 0x40;
+
+/// Disable IRQ interrupts to avoid getting stuck in a loop while single stepping
+///
+/// When GDB hits a single breakpoint and resumes it will follow the steps:
+///  - Clear SW breakpoint we've hit
+///  - Single step
+///  - Re-insert the SW breakpoint
+///  - Resume
+/// However, with IRQ enabled the single step takes us into the IRQ handler so when we resume we
+/// immediately hit the SW breapoint we just re-inserted getting stuck in a loop.
+fn toggle_interrupts(vcpu_fd: &VcpuFd, enable: bool) -> Result<(), GdbTargetError> {
+    let kreg_off = offset_of!(user_pt_regs, pstate);
+    let pstate_id = arm64_core_reg_id!(KVM_REG_SIZE_U64, kreg_off);
```

Comment:
This can be defined as a constant somewhere at the top of the file.

Reply:
Moved to a constant now